src/Decompiler/Evaluation/ComplicatedNotRule.cs

```diff
@@ -0,0 +1,86 @@
+#region License
+/*
+ * Copyright (C) 2020 Sven Almgren.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
+ */
+#endregion
+
+using Reko.Core;
+using Reko.Core.Expressions;
+using Reko.Core.Operators;
+using Reko.Core.Types;
+using System;
+
+namespace Reko.Evaluation
+{
+    /// <summary>
+    /// Rule that matches (0 - (-EXPRESSION == 0) + 1) and generates (!EXPRESSION).
+    /// This is a solution for a very specific sequence of instructions.
+    ///
+    /// Example x86 assembler:
+    /// neg ax
+    /// sbb ax, ax
+    /// inc ax
+    ///
+    /// Example result:
+    /// turn func(0x00 - (-MyVariable == 0x00) + 0x01) into func(!MyVariable)
+    /// </summary>
+    public class ComplicatedNotRule
```

Comment:
The naming is an issue; it's conceivable even more complicated `Not` rules might be needed. Naming is hard. How about `LogicalNotFromAdditionRule`? The initial expression is forming a logical not from the additions.

Reply:
Sure thing! I'll look into it later.

Reply:
Unless you say something else I'm going with `LogicalNotFromBorrowRule` as that's what I expect the majority of sources will be. Other platforms can probably create something else, but unless it's done with the borrow-magic I would guess they produce some other formula.

You can give suggestions here, or on next push :P But I think that this name is more informative.

Reply:
I'm fine with that name.